chronos/asyncfutures2.nim

```diff
@@ -284,6 +287,9 @@ proc addCallback*(future: FutureBase, cb: CallbackFunc, udata: pointer = nil) =
   ## Adds the callbacks proc to be called when the future completes.
   ##
   ## If future has already completed then ``cb`` will be called immediately.
+  {.gcsafe.}:
```

Comment:
not sure if chronos is thread safe, but this is certainly not

also, it seems wrong to be creating strings here and counting the callbacks like this - normally what's done is that if a particular counter is expensive, you count for example 1% of the instances and scale the result accordingly, getting a statistical approximation - something that the metrics library should take care of - if metrics require you to pre-aggregate stuff, something is wrong with the library.

finally, `when metrics` since it's not necessarily part of every codebase to be running this stuff

Reply:
I don't want all future types ending up in the metrics. Just the most frequently used ones, hence the aggregation, sorting and truncating (further down, in the other file). I know it's ugly, but dragging around hundreds of labels and trying to sort them with Prometheus queries is uglier.

If I add `when defined(metrics):` all over the code, it will get less readable. Is the performance gain worth it?

No, it's not thread-safe. When we start using multiple event loops in parallel and we want combined metrics, we'll need locks.

Reply:
It's not just about runtime performance but also compile time and dependency hygiene.. 

also, like this, there's a significant chunk of metrics code intertwined with chronos' callback / future logic - code-wise, the metrics stuff should be factored out into its own functions at least so that it takes up less code-space where the important things are going on - it's already a readability issue that needs to be fixed